groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -218,16 +219,243 @@ extern const Optional_OptNoSuchType optNoSuchType;
     // functions that are constrained using a function argument.
 
 #ifdef BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
+
+#define BSLSTL_OPTIONAL_OR_IS_CONSTRUCTIBLE_V(U, V)                           \
+    || std::is_constructible<U, V>::value
+#define BSLSTL_OPTIONAL_AND_IS_CONSTRUCTIBLE_V(U, V)                          \
+    &&std::is_constructible<U, V>::value
+#define BSLSTL_OPTIONAL_AND_IS_ASSIGNABLE_V(U, V)                             \
+    &&std::is_assignable<U, V>::value
 #define BSLSTL_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE                             \
     std::is_trivially_destructible
 
 #else
+
+#define BSLSTL_OPTIONAL_OR_IS_CONSTRUCTIBLE_V(U, V)
+#define BSLSTL_OPTIONAL_AND_IS_CONSTRUCTIBLE_V(U, V)
+#define BSLSTL_OPTIONAL_AND_IS_ASSIGNABLE_V(U, V)
 #define BSLSTL_OPTIONAL_IS_TRIVIALLY_DESTRUCTIBLE bsl::is_trivially_copyable
     // C++03 does not provide a trivially destructible trait.  Instead we use
     // 'bsl::is_trivially_copyable' which implies the type is also trivially
     // destructible.
 #endif  // BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY
 
+// Type traits to assist in choosing the correct assignment and construction
+// overload.  If the 'value_type' converts  or assigns from an
+// 'optional<other_type>', then the overload passing the function parameter to
+// the 'value_type' is preferred.  As in 'std' implementation, if the
+// 'value_type' converts or assigns from any value category, we consider it
+// convertible/assignable from optional.
+
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+template <class TYPE, class ANY_TYPE>
+struct Optional_ConvertsFromStdOptional
+: bsl::integral_constant<
+      bool,
+      bsl::is_convertible<const std::optional<ANY_TYPE>&, TYPE>::value ||
+          bsl::is_convertible<std::optional<ANY_TYPE>&, TYPE>::value ||
+          bsl::is_convertible<const std::optional<ANY_TYPE>, TYPE>::value ||
+          bsl::is_convertible<std::optional<ANY_TYPE>, TYPE>::value ||
+          std::is_constructible<TYPE, const std::optional<ANY_TYPE>&>::value ||
+          std::is_constructible<TYPE, std::optional<ANY_TYPE>&>::value ||
+          std::is_constructible<TYPE, const std::optional<ANY_TYPE> >::value ||
+          std::is_constructible<TYPE, std::optional<ANY_TYPE> >::value> {
```

Comment:
We usually place the operators in multiline boolean expressions like this:

```suggestion
    bsl::is_convertible<const std::optional<ANY_TYPE>&, TYPE>::value
        || bsl::is_convertible<      std::optional<ANY_TYPE>&, TYPE>::value
        || bsl::is_convertible<const std::optional<ANY_TYPE>,  TYPE>::value
        || bsl::is_convertible<      std::optional<ANY_TYPE>,  TYPE>::value
        || std::is_constructible<TYPE, const std::optional<ANY_TYPE>&>::value
        || std::is_constructible<TYPE,       std::optional<ANY_TYPE>&>::value
        || std::is_constructible<TYPE, const std::optional<ANY_TYPE> >::value
        || std::is_constructible<TYPE,       std::optional<ANY_TYPE> >::value>
{
```

This is manual though :/  Perhaps we should ask Mike if he thinks if this is important enough.

Reply:
I don't mind changing it, however this style is imposed by bde-format. If I change it manually, it will revert back the changes next time I run it.  Do you still want me to change it ?

Reply:
bde-format unfortunately is not perfect in many instances.  I think that it would make sense to leave it as bde-format suggests now, and perhaps revisit this formatting when the PR is basically ready to be merged.